Make export var or export *'s with duplicate identifiers an error

weswigham opened the pull request:
Closes #4527. The relevant spec points were referenced in that issue. Namespaces, interfaces, and function overloads (so TS constructs) are exempt from the duplicate identifier check.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1020,38 +1020,62 @@ namespace ts {
+                            );
+                        }
+                        for (let id in lookupTable) {
+                            if (id !== "export=" && lookupTable[id].exportsWithDuplicate.length && !(id in symbols)) { // It's not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
```
Move the comment onto a line above

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1020,38 +1020,62 @@ namespace ts {
+                        for (let id in lookupTable) {
+                            if (id !== "export=" && lookupTable[id].exportsWithDuplicate.length && !(id in symbols)) { // It's not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
+                                for (let node of lookupTable[id].exportsWithDuplicate) {
+                                    diagnostics.add(createDiagnosticForNode(node, Diagnostics.An_export_Asterisk_from_0_declaration_has_already_exported_a_member_named_1_Consider_explicitly_re_exporting_to_resolve_the_ambiguity, lookupTable[id].specifierText, id));
```
Break this into multiple lines

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13698,13 +13722,33 @@ namespace ts {
         function checkExternalModuleExports(node: SourceFile | ModuleDeclaration) {
             let moduleSymbol = getSymbolOfNode(node);
-            let links = getSymbolLinks(moduleSymbol);
+            let links: SymbolLinks = getSymbolLinks(moduleSymbol);
```
Why did you need this type annotation?

weswigham replied:
I don't. :3 I wrote it in to use go to definition on it, and forgot to remove it.

DanielRosenwasser replied:
https://github.com/dotnet/roslyn/issues/2505

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13698,13 +13722,33 @@ namespace ts {
+                let exports = getExportsOfModule(moduleSymbol); // Checks for export * conflicts
+                for (let id in exports) {
+                    if (id === "__export") continue;
+                    if (!(exports[id].flags & SymbolFlags.Namespace || exports[id].flags & SymbolFlags.Interface) && exports[id].declarations.length > 1) { // 15.2.1.1 It is a Syntax Error if the ExportedNames of ModuleItemList contains any duplicate entries. (TS Exceptions: namespaces, function overloads, and interfaces)
```
Move the comment above

DanielRosenwasser replied:
What about enums?

DanielRosenwasser replied:
Extract `exports[id]` to a variable so you avoid multiple lookups on `id`

weswigham replied:
Good point about enums - they do merge, don't they. I'll add them to the conditional and to the test.

weswigham commented:
@DanielRosenwasser updated.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1020,38 +1020,68 @@ namespace ts {
         }
 
-        function extendExportSymbols(target: SymbolTable, source: SymbolTable) {
+        interface ExportStarDiagnosticsLookupTable {
```
Why not just make this a `Map<ExportDeclaration[]>` and use the `id` as the text?

weswigham replied:
That's fine, but naming the type yields more informative documentation, no?

weswigham replied:
Also, the text isn't the id of the symbol in the map - the map key id is the duplicated symbol's id, the text is the text of the export specifier it was first found in.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1020,38 +1020,68 @@ namespace ts {
                     target[id] = source[id];
+                    if (lookupTable && exportNode) {
+                        lookupTable[id] = {
+                            specifierText: getTextOfNode(exportNode.moduleSpecifier),
```
I'd prefer to use the`.text` because the message will look awkward if you use single quotes.

weswigham replied:
...What if I change the message to use backticks around the code quote, markdown style? I'd rather not change the user's text on the import in the error where possible - makes it less ctrl-f'able.

DanielRosenwasser replied:
Maybe, though that always looks odd somehow. Bring it up with @mhegazy.

vladima replied:
since `getTextOfNode(exportNode.moduleSpecifier)` is the same on every iteration can we get it once when it is necessary and then reuse?

weswigham commented:
@mhegazy I've refreshed this PR  - we'd probably still want it so we align with the spec, right?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13978,6 +14009,28 @@ namespace ts {
                     const declaration = getDeclarationOfAliasSymbol(exportEqualsSymbol) || exportEqualsSymbol.valueDeclaration;
                     error(declaration, Diagnostics.An_export_assignment_cannot_be_used_in_a_module_with_other_exported_elements);
                 }
+                const exports = getExportsOfModule(moduleSymbol); // Checks for export * conflicts
```
Move comment above

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13978,6 +14009,28 @@ namespace ts {
                 }
+                const exports = getExportsOfModule(moduleSymbol); // Checks for export * conflicts
+                for (const id in exports) {
+                    if (id === "__export") continue;
```
Curlies

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1055,38 +1055,69 @@ namespace ts {
+                        }
+                        for (const id in lookupTable) {
+                            // It's not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
+                            if (id !== "export=" && lookupTable[id].exportsWithDuplicate.length && !(id in symbols)) {
```
Instead of nesting the `for`, negate the condition and bail out

DanielRosenwasser replied:
Also, declare a variable for `lookupTable[id].exportsWithDuplicate`

weswigham commented:
@DanielRosenwasser feedback incorporated

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,71 @@ namespace ts {
+                        }
+                        for (const id in lookupTable) {
+                            const { exportsWithDuplicate } = lookupTable[id];
+                            // It's not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
```
no apostrophe

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -14040,8 +14073,29 @@ namespace ts {
+                        continue;
+                    }
+                    const exportedSymbol = exports[id];
+                     // 15.2.1.1 It is a Syntax Error if the ExportedNames of ModuleItemList contains any duplicate entries. (TS Exceptions: namespaces, function overloads, enums, and interfaces)
```
Prefix this with the spec you're quoting (i.e. TS or ECMA262)

DanielRosenwasser replied:
Should we be allowing these exceptions?

weswigham replied:
Based on the way we do should ES6 exports for namespaces and enums, I believe so (the others are types and are fine). Namespaces and enums should end up roughly like this in ES6 forms:
```js
var Namespace_1;
(function(Namespace_1){
 // ...
})(Namespace_1 = Namespace_1 || {});
var Namespace_1;
(function(Namespace_1){
 // ...
})(Namespace_1 = Namespace_1 || {});
export var Namespace = Namespace_1;
```

*checks*
Our present emit is actually
```js
export var Foo;
(function (Foo) {
    Foo.cat = 2;
})(Foo || (Foo = {}));
export var Foo;
(function (Foo) {
    Foo.dog = 2;
})(Foo || (Foo = {}));
```
Which is in this case incorrect (it introduces duplicated exported identifiers) - I'll need to change the emit.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -14040,8 +14073,29 @@ namespace ts {
+                    }
+                    const exportedSymbol = exports[id];
+                     // 15.2.1.1 It is a Syntax Error if the ExportedNames of ModuleItemList contains any duplicate entries. (TS Exceptions: namespaces, function overloads, enums, and interfaces)
+                    if (!(exportedSymbol.flags & SymbolFlags.Namespace || exportedSymbol.flags & SymbolFlags.Interface || exportedSymbol.flags & SymbolFlags.Enum) && exportedSymbol.declarations.length > 1) {
```
`exportedSymbol.flags & (SymbolFlags.NameSpace | SymbolFlags.Interface | SymbolFlags.Enum)`

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -14040,8 +14073,29 @@ namespace ts {
+                    if (id === "__export") {
+                        continue;
+                    }
+                    const exportedSymbol = exports[id];
```
Consider destructuring `flags` and `declarations`

weswigham commented:
@DanielRosenwasser I've incorporated your feedback and in response to your feedback about weather namespaces should be exempt or not and corrected their emit - I still need to check enums, though.

weswigham commented:
Handled enums correctly - now enums and namespaces will have only one exported binding when targeting ES6 modules, no matter how many declarations they have. :+1:

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -14040,8 +14073,29 @@ namespace ts {
             }
+
+            function isNotOverload(declaration: Declaration): boolean {
+                return (declaration.kind !== SyntaxKind.FunctionDeclaration || typeof (declaration as FunctionDeclaration).body !== "undefined");
```
Remove parens - also do you need to use `typeof`?

weswigham replied:
...probably not. For some reason I thought body was a `string`, when it should be `FunctionBody`.

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -836,6 +836,10 @@
         "category": "Error",
         "code": 2307
     },
+    "An `export * from {0}` declaration has already exported a member named '{1}'. Consider explicitly re-exporting to resolve the ambiguity.": {
```
Module {0} has already exported a member named '{1}'. Consider explicitly re-exporting to resolve the ambiguity."

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,71 @@ namespace ts {
+            exportsWithDuplicate: ExportDeclaration[];
+        }
+
+        function extendExportSymbols(target: SymbolTable, source: SymbolTable, lookupTable?: Map<ExportedSymbolDiagnostics>, exportNode?: ExportDeclaration) {
```
At a cursory glance, it's hard for me to understand exactly what's going on. Can you add a quick comment for what this is and what it's used for?

For instance, what does `lookupTable` do? Which of these are you going to modify?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,71 @@ namespace ts {
         }
 
-        function extendExportSymbols(target: SymbolTable, source: SymbolTable) {
+        interface ExportedSymbolDiagnostics {
```
It's not clear what this is; "diagnostics" typically implies something to do with error messages.

weswigham replied:
It's an structure used just to collect data to generate an error message from, so Diagnostics seems appropriate.

DanielRosenwasser replied:
What about `ExportCollisionTracker`?

weswigham replied:
Sure.

weswigham commented:
@DanielRosenwasser any further comments?

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -6209,9 +6209,11 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
-                    if (!(node.flags & NodeFlags.Export) || isES6ExportedDeclaration(node)) {
+
+                    const isES6ExportedEnum = isES6ExportedDeclaration(node);
+                    if ((!(node.flags & NodeFlags.Export)) || (isES6ExportedEnum && isFirstDeclarationOfKind(node, node.symbol && node.symbol.declarations, SyntaxKind.EnumDeclaration))) {
```
I think there's an extra nested paren

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -6321,15 +6327,18 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
-                    if (isES6ExportedDeclaration(node)) {
-                        write("export ");
+                    const isES6ExportedNamespace = isES6ExportedDeclaration(node);
+                    if ((!isES6ExportedNamespace) || isFirstDeclarationOfKind(node, node.symbol && node.symbol.declarations, SyntaxKind.ModuleDeclaration)) {
```
Is it appropriate to assume that a node has a a symbol? Maybe just `Debug.assert(!!node.symbol)`

DanielRosenwasser replied:
Remvoe extra paren

DanielRosenwasser commented:
:+1:

weswigham commented:
Ran RWC tests for this change and found it only affected angular - _however_ it exposed a bug in my check, it is _valid_ to reexport the same thing via multiple `export *`'s provided those bindings are to the same value within the same module (eg, module A reexports foo from C and foo from B, while module B is reexporting foo from C) - this is okay because the resolution is not "ambiguous", as both bindings are the same value from the same module. (ECMA 262 [15.2.1.16.3](http://www.ecma-international.org/ecma-262/6.0/#sec-resolveexport) 10.f.ii.2) Once this is accounted for in this PR, I don't think angular (or anything else in our RWC tests) will have any errors with this change.

weswigham commented:
Now all rwc tests pass with this change. :+1:

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -14095,8 +14132,29 @@ namespace ts {
+                    if (id === "__export") {
+                        continue;
+                    }
+                    const {declarations, flags} = exports[id];
```
Space inside of `{ }`

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
+                        }
+                        for (const id in lookupTable) {
+                            const { exportsWithDuplicate } = lookupTable[id];
+                            // Its not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
```
It's

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
-                        }
-                        extendExportSymbols(result, symbol.exports);
-                    }
+                    const symbols: SymbolTable = cloneSymbolTable(symbol.exports);
```
Do you need this type annotation?

weswigham replied:
Probably not.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
+                        for (const id in lookupTable) {
+                            const { exportsWithDuplicate } = lookupTable[id];
+                            // Its not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
+                            if (id === "export=" || !exportsWithDuplicate.length || id in symbols) {
```
Use `hasProperty`

weswigham replied:
Is this a rule for us?

DanielRosenwasser replied:
Maybe it should be, but there are places where we probably use `in` *in*tentionally

weswigham replied:
:clap:

DanielRosenwasser commented:
:+1:

weswigham commented:
@vladima would you go over this PR?

vladima commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
+        }
+
+        /**
+         * Extends one symbol table with abother while collecting information on name collisions for error message generation into the `lookupTable` argument
```
another

vladima commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
+                        };
+                    }
+                }
+                else if (lookupTable && exportNode && id !== "default" && hasProperty(target, id) && resolveSymbol(target[id]) !== resolveSymbol(source[id])) {
```
feels like the same condition ` id !== "default" && hasProperty(target, id)` is already checked above. Can you lift it to a variable?

vladima commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
             // module 'a' can 'export * from "b"' and 'b' can 'export * from "a"' without error.
-            function visit(symbol: Symbol) {
+            function visit(symbol: Symbol): SymbolTable {
                 if (symbol && symbol.flags & SymbolFlags.HasExports && !contains(visitedSymbols, symbol)) {
```
can this condition be inverted to reduce nesting?

vladima commented on src/compiler/checker.ts:
```diff
@@ -1093,38 +1093,75 @@ namespace ts {
+                    if (lookupTable && exportNode) {
+                        lookupTable[id] = {
+                            specifierText: getTextOfNode(exportNode.moduleSpecifier),
+                            exportsWithDuplicate: []
```
can we defer array allocation until it is actually needed?
